File: cache_interceptor/src/main/java/uk/co/glass_software/android/cache_interceptor/interceptors/error/ApiError.java
package uk.co.glass_software.android.cache_interceptor.interceptors.error;

import android.support.annotation.Nullable;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.apache.commons.lang3.builder.ToStringBuilder;

import uk.co.glass_software.android.cache_interceptor.utils.Function;

public class ApiError extends Exception implements Function<ApiError, Boolean> {
    
    public final static int NON_HTTP_STATUS = -1;
    
    private final Throwable throwable;
    private final int httpStatus;
    private final ErrorCode errorCode;
    @Nullable private final String description;
    
    public ApiError(Throwable throwable) {
        this(throwable,
             NON_HTTP_STATUS,
             ErrorCode.UNKNOWN,
             null
        );
    }
    
    public ApiError(Throwable throwable,
                    int httpStatus,
                    ErrorCode errorCode,
                    @Nullable String description) {
        this.throwable = throwable;
        this.httpStatus = httpStatus;
        this.errorCode = errorCode;
        this.description = description;
    }
    
    @Override //isNetworkError
    public Boolean get(ApiError apiError) {
        return apiError.isNetworkError();
    }
    
    public int getHttpStatus() {
        return httpStatus;
    }
    
    public ErrorCode getErrorCode() {
        return errorCode;
    }
    
    @Nullable
    public String getDescription() {
        return description;
    }
    
    public Boolean isNetworkError() {
        return errorCode == ErrorCode.NETWORK;
    }
    
    @Nullable
    public static ApiError from(Throwable throwable) {
        if (throwable instanceof ApiError) {
            return (ApiError) throwable;
        }
        return null;
    }
    
    public static boolean isNetworkError(Throwable throwable) {
        ApiError apiError = from(throwable);
        return apiError != null && apiError.isNetworkError();
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        
        ApiError apiError = (ApiError) o;
        
        return new EqualsBuilder()
                .append(httpStatus, apiError.httpStatus)
                .append(throwable, apiError.throwable)
                .append(errorCode, apiError.errorCode)
                .append(description, apiError.description)
                .isEquals();
    }
    
    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .append(throwable)
                .append(httpStatus)
                .append(errorCode)
                .append(description)
                .toHashCode();
    }
    
    @Override
    public String toString() {
        return new ToStringBuilder(this)
                .append("throwable", throwable)
                .append("httpStatus", httpStatus)
                .append("errorCode", errorCode)
                .append("description", description)
                .toString();
    }
}

File: cache_interceptor/src/main/java/uk/co/glass_software/android/dejavu/interceptors/internal/error/ApiError.kt
/*
 * Copyright (C) 2017 Glass Software Ltd
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package uk.co.glass_software.android.dejavu.interceptors.internal.error

import uk.co.glass_software.android.dejavu.configuration.NetworkErrorProvider
import uk.co.glass_software.android.dejavu.interceptors.internal.error.ErrorCode.UNKNOWN

data class ApiError constructor(private val throwable: Throwable,
                                val httpStatus: Int = NON_HTTP_STATUS,
                                val errorCode: ErrorCode = UNKNOWN,
                                val description: String? = null)
    : Exception(throwable),
        NetworkErrorProvider {

    override fun isNetworkError() = errorCode === ErrorCode.NETWORK

    companion object {
        const val NON_HTTP_STATUS = -1
        fun from(throwable: Throwable) = throwable as? ApiError
    }
}